In [1]:
import pandas as pd

# Load the weather dataset
weather_data = pd.read_excel('/Users/samritha/Downloads/final_weather_data.xlsx', header=1)

print(weather_data.head())

       States  2007  2008  2009  2010  2011  2012  2013  2014  2015  2016  \
0     Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  43.0   
1      Alaska   5.5   0.4  -0.2   1.5   7.0 -12.8   9.8  18.4   7.1  17.2   
2     Arizona  39.0  39.8  45.2  42.2  41.5  45.3  38.6  46.5  45.1  41.8   
3    Arkansas  39.9  39.0  38.4  36.5  37.6  44.6  42.1  35.2  38.6  39.5   
4  California  41.5  41.1  46.9  45.2  45.2  46.5  42.1  50.1  48.6  44.5   

   2017  2018  2019  2020  2021  2022  2023  
0  52.4  40.4  46.5  49.5  46.6  43.9  51.4  
1   4.5   6.7   7.5  -6.4  10.6   2.6  10.9  
2  42.3  47.6  42.4  43.7  42.6  43.5  40.2  
3  45.2  35.7  40.8  43.9  41.4  38.7  46.4  
4  42.2  48.9  46.1  45.3  45.5  46.2  42.6  


In [2]:
# List of all 50 U.S. states
all_states = [
    "Alabama", "Alaska", "Arizona", "Arkansas", "California", "Colorado", 
    "Connecticut", "Delaware", "Florida", "Georgia", "Hawaii", "Idaho", 
    "Illinois", "Indiana", "Iowa", "Kansas", "Kentucky", "Louisiana", 
    "Maine", "Maryland", "Massachusetts", "Michigan", "Minnesota", 
    "Mississippi", "Missouri", "Montana", "Nebraska", "Nevada", 
    "New Hampshire", "New Jersey", "New Mexico", "New York", 
    "North Carolina", "North Dakota", "Ohio", "Oklahoma", "Oregon", 
    "Pennsylvania", "Rhode Island", "South Carolina", "South Dakota", 
    "Tennessee", "Texas", "Utah", "Vermont", "Virginia", "Washington", 
    "West Virginia", "Wisconsin", "Wyoming"
]

# Clean up state names in the dataset by removing extra spaces and standardizing capitalization
weather_data['States'] = weather_data['States'].str.strip().str.title()

# Get the unique states from the dataset
dataset_states = weather_data['States'].unique()

# Find the missing states by comparing the two lists
missing_states = set(all_states) - set(dataset_states)

print("Missing states:", missing_states)

Missing states: {'Hawaii'}


In [3]:
import pandas as pd

# Load the ACS data
acs_data = pd.read_csv('/Users/samritha/Downloads/acs_data.csv', low_memory=False)

print(acs_data.shape)

(1093982, 13)


In [4]:
# Extract the two-digit FIPS code from the geo_id. Here, we assume it's the 10th and 11th characters.
acs_data['state_fips'] = acs_data['geo_id'].str[9:11]

# Map the FIPS codes to state initials
fips_to_initials = {
    "01": "AL", "02": "AK", "04": "AZ", "05": "AR", "06": "CA", "08": "CO",
    "09": "CT", "10": "DE", "12": "FL", "13": "GA", "15": "HI", "16": "ID",
    "17": "IL", "18": "IN", "19": "IA", "20": "KS", "21": "KY", "22": "LA",
    "23": "ME", "24": "MD", "25": "MA", "26": "MI", "27": "MN", "28": "MS",
    "29": "MO", "30": "MT", "31": "NE", "32": "NV", "33": "NH", "34": "NJ",
    "35": "NM", "36": "NY", "37": "NC", "38": "ND", "39": "OH", "40": "OK",
    "41": "OR", "42": "PA", "44": "RI", "45": "SC", "46": "SD", "47": "TN",
    "48": "TX", "49": "UT", "50": "VT", "51": "VA", "53": "WA", "54": "WV",
    "55": "WI", "56": "WY"
}

# Use the state FIPS codes to add a new column with state initials
acs_data['state_initials'] = acs_data['state_fips'].map(fips_to_initials)

print(acs_data[['geo_id', 'state_initials']].head())

                 geo_id state_initials
0             Geography            NaN
1  1400000US01001020100             AL
2  1400000US01001020100             AL
3  1400000US01001020100             AL
4  1400000US01001020100             AL


In [5]:
# Create a dictionary to map state names to FIPS codes
state_to_fips = {
    "Alabama": "01", "Alaska": "02", "Arizona": "04", "Arkansas": "05", "California": "06",
    "Colorado": "08", "Connecticut": "09", "Delaware": "10", "Florida": "12", "Georgia": "13",
    "Hawaii": "15", "Idaho": "16", "Illinois": "17", "Indiana": "18", "Iowa": "19", "Kansas": "20",
    "Kentucky": "21", "Louisiana": "22", "Maine": "23", "Maryland": "24", "Massachusetts": "25",
    "Michigan": "26", "Minnesota": "27", "Mississippi": "28", "Missouri": "29", "Montana": "30",
    "Nebraska": "31", "Nevada": "32", "New Hampshire": "33", "New Jersey": "34", "New Mexico": "35",
    "New York": "36", "North Carolina": "37", "North Dakota": "38", "Ohio": "39", "Oklahoma": "40",
    "Oregon": "41", "Pennsylvania": "42", "Rhode Island": "44", "South Carolina": "45", "South Dakota": "46",
    "Tennessee": "47", "Texas": "48", "Utah": "49", "Vermont": "50", "Virginia": "51", "Washington": "53",
    "West Virginia": "54", "Wisconsin": "55", "Wyoming": "56"
}

# Add a 'state_fips' column to the weather_data by mapping 'States'
weather_data['state_fips'] = weather_data['States'].map(state_to_fips)

# Now you can merge on 'state_fips'
merged_data = weather_data.merge(acs_data[['state_fips', 'state_initials', 'geo_id']], on='state_fips', how='left')

# Display the merged data to verify
print(merged_data.head())
print(merged_data.tail())

# Display the Shape of merged data
print(merged_data.shape)

    States  2007  2008  2009  2010  2011  2012  2013  2014  2015  ...  2017  \
0  Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  ...  52.4   
1  Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  ...  52.4   
2  Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  ...  52.4   
3  Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  ...  52.4   
4  Alabama  46.7  43.0  45.3  40.0  41.1  50.1  50.0  37.3  43.6  ...  52.4   

   2018  2019  2020  2021  2022  2023  state_fips state_initials  \
0  40.4  46.5  49.5  46.6  43.9  51.4          01             AL   
1  40.4  46.5  49.5  46.6  43.9  51.4          01             AL   
2  40.4  46.5  49.5  46.6  43.9  51.4          01             AL   
3  40.4  46.5  49.5  46.6  43.9  51.4          01             AL   
4  40.4  46.5  49.5  46.6  43.9  51.4          01             AL   

                 geo_id  
0  1400000US01001020100  
1  1400000US01001020100  
2  1400000US01001020100  
3  1400000US

In [6]:
# Check for missing values in merged_data
print(merged_data.isnull().sum())

States            0
2007              0
2008              0
2009              0
2010              0
2011              0
2012              0
2013              0
2014              0
2015              0
2016              0
2017              0
2018              0
2019              0
2020              0
2021              0
2022              0
2023              0
state_fips        0
state_initials    0
geo_id            0
dtype: int64


In [7]:
# Define the desired column order
column_order = ['States', 'state_fips', 'state_initials', 'geo_id',
                2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 
                2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]

# Rearrange the columns in the specified order
merged_data = merged_data[column_order]

# Rename the columns
merged_data = merged_data.rename(columns={
    'state_fips': 'State_No.',
    'state_initials': 'State_Initials',
    'geo_id': 'Geo_Id'
})

# Display the first few rows to confirm the new column order
print(merged_data.head())


    States State_No. State_Initials                Geo_Id  2007  2008  2009  \
0  Alabama        01             AL  1400000US01001020100  46.7  43.0  45.3   
1  Alabama        01             AL  1400000US01001020100  46.7  43.0  45.3   
2  Alabama        01             AL  1400000US01001020100  46.7  43.0  45.3   
3  Alabama        01             AL  1400000US01001020100  46.7  43.0  45.3   
4  Alabama        01             AL  1400000US01001020100  46.7  43.0  45.3   

   2010  2011  2012  ...  2014  2015  2016  2017  2018  2019  2020  2021  \
0  40.0  41.1  50.1  ...  37.3  43.6  43.0  52.4  40.4  46.5  49.5  46.6   
1  40.0  41.1  50.1  ...  37.3  43.6  43.0  52.4  40.4  46.5  49.5  46.6   
2  40.0  41.1  50.1  ...  37.3  43.6  43.0  52.4  40.4  46.5  49.5  46.6   
3  40.0  41.1  50.1  ...  37.3  43.6  43.0  52.4  40.4  46.5  49.5  46.6   
4  40.0  41.1  50.1  ...  37.3  43.6  43.0  52.4  40.4  46.5  49.5  46.6   

   2022  2023  
0  43.9  51.4  
1  43.9  51.4  
2  43.9  51.4  
3  4

In [8]:
# Save the merged dataset 
merged_data.to_csv('merged_weather_data.csv', index=False)

### Key Insights
#### ACS data shape: (1093982, 13)
#### Missing State in Weather data: Hawaii 
#### After merging Weather data with Geo_id shape: (1072433, 21)
#### Rows Missing in Merged data: 21549 rows